File: algorithms/searching/binary-search/go/binary_search.go
package main

import (
	"fmt"
)

// BinarySearch searches for the target in a sorted array and returns its index or -1 if not found.
func BinarySearch(arr []int, target int) int {
	low, high := 0, len(arr)-1

	for low <= high {
		mid := (low + high) / 2
		if arr[mid] == target {
			return mid
		} else if arr[mid] < target {
			low = mid + 1
		} else {
			high = mid - 1
		}
	}

	return -1
}

func main() {
	tests := []struct {
		arr      []int
		target   int
		expected int
	}{
		{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 5, 4},
		{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 11, -1},
		{[]int{1, 3, 5, 7, 9}, 3, 1},
		{[]int{1, 3, 5, 7, 9}, 2, -1},
	}

	for _, test := range tests {
		result := BinarySearch(test.arr, test.target)
		if result != test.expected {
			fmt.Printf("Test failed for array %v with target %d. Expected %d, got %d.\n", test.arr, test.target, test.expected, result)
			return
		}
	}

	fmt.Println("All tests passed!")
}

File: algorithms/searching/binary-search/binary_search.py
def binary_search(arr, target):
    """
    A function to implement the binary search algorithm.

    Parameters:
    - arr: A sorted list of items.
    - target: The item to search for in the list.

    Returns:
    - int: The index of the target if found, otherwise -1.
    """
    # Initialize the low and high pointers.
    low, high = 0, len(arr) - 1

    # Loop until the low and high pointers meet.
    while low <= high:
        # while low < high:
        mid = (low + high) // 2

        # Check if the target is at the middle.
        if arr[mid] == target:
            # Target found. Return the index.
            return mid
        elif arr[mid] < target:
            # Target lies to the right of mid. Move the low pointer.
            low = mid + 1
        else:
            # Target lies to the left of mid. Move the high pointer.
            high = mid - 1

    # Target not found. Return -1.
    return -1

# Driver code.
if __name__ == "__main__":
    # Get the input from the user.
    arr = [
        int(x)
        for x in input("Enter a sorted list of numbers separated by spaces: ").split()
    ]
    # Get the target from the user.
    target = int(input("Enter the number you want to search for: "))
    # Call the binary search function.
    result = binary_search(arr, target)
    # Print the result.
    if result != -1:
        print(f"Element {target} is present at index {result}.")
    else:
        print(f"Element {target} is not present in the array.")
